### ISSUE TITLE
Question: where to place winrt/uwp questions?

### ISSUE THREAD

**Issue by jtorjo** [microsoft/microsoft-ui-xaml#1700]
WPF has its own repository on github. Toolkit - same. WinUI - same.

Winrt / UWP - not so much.
It's really weird that Microsoft wants us (or does it?) to use UWP , but there's no place to ask questions about it.

No, stackoverflow is not the place to ask advanced questions about winrt/uwp.
No, Feedback Hub is not the place to ask questions about winrt/uwp - not sure if anyone reads it.
No, the new UWP forum is not the place either - simple questions, sure, but anything advanced, and boom - no answer, or just answers avoiding the question (case in point - https://docs.microsoft.com/answers/questions/608/please-either-allow-systemio-all-over-hdd-or-drast.html).

Also, I'm curious about where both Winrt/UWP are heading - is there **any** development on them? Are you listening to **any** feedback from developers trying to use it?

Winrt seems to have started deeply flawed, and it's remained this way.

Here are some questions that are haunting my nights:
1. [winrt] Storage folder queries (https://docs.microsoft.com/answers/questions/608/please-either-allow-systemio-all-over-hdd-or-drast.html) - this is something known FOR YEARS, yet no one is even trying to fix it.

2. [winrt] LoadAsync - sure, great, now you've created an awesome way to load stuff async. Awesome! But I want the ability to load stuff synchronously. Maybe I have an in-memory stream, whatever. I want to load stuff synchronously!

3. [winrt] Loading bitmaps **only on the main thread, and async**. This is simply beyond my understanding. Why why why would you ENFORCE that? I want to load bitmaps on any thread, and from memory streams, and i want this to happen FAST. This is simply not possible. I'm porting code from my former WPF app, and it's simply impossible to make it work efficient enough.

These are just a few questions. I could probably write a book, and I'm sure other people have lots and lots of questions/ suggestions for improvement.
We'd just want a place to ask.

Best,
John

**Comment by Felix-Dev** [comment]
> Also, I'm curious about where both Winrt/UWP are heading - is there any development on them? Are you listening to any feedback from developers trying to use it?

WinUI is based on the UWP UIs APIs so any development you are currently seeing in this repo is still "UWP development". It's just decoupled from the rest of the UWP APIs.
With every Windows 10 SDK release, the Windows Runtime will get new and improved APIs. Those SDK releases might not contain the changes you want to see but the Windows Runtime is clearly in active development.

As to where to list your questions: Anything regarding the UI APIs of the UWP (Xaml, Composition, Input, Windowing,...): This is the place to ask your questions!

For the remaining UWP APIs: Feedback Hub is currently the official place to report problems/suggest features. For general questions, Microsoft Q&A seems to be the official platform. Q&A platform is still "new" so perhaps give it some more time. 

I can relate to your frustration though. All I can say in addition is that MS in general at least seems to be aware of the frustration regarding the issues/feedback reporting system. And while I don't know if we'll see improvements for non-WinUI UWP APIs anytime soon, at least the WinUI team is doing a good job dealing with reported issues & suggestions.

**Comment by jtorjo** [comment]
> With every Windows 10 SDK release, the Windows Runtime will get new and improved APIs. Those SDK releases might not always contain the changes you want to see but the Windows Runtime is clearly in active development.

I understand, but what useful stuff has happened there (in WinRT)? For us developers, that is?
How is Microsoft listening for our voice, if there's pretty much no place for us to actually have one?

**I simply don't understand how I can implement any bitmap caching strategy, with the crappy API that comes from WinRT.**

> 
> As to where to list your questions: Anything regarding the UI APIs of the UWP (Xaml, Composition, Input, Windowing,...): This is the place to ask your questions/report issues/suggest features!

Got it, thanks!
 
> I can relate to your frustration though. All I can say in addition is that MS in general at least seems to be aware of the frustration regarding the issues/feedback reporting system. And while I don't know if we'll see improvements to the feedback system for non-WinUI UWP APIs anytime soon, at least the WinUI team is doing a good job dealing with reported issues & suggestions.

Agree 100% about the WinUI team - it is doing an awesome job!

**Comment by Gavin-Williams** [comment]
"[winrt] Loading bitmaps only on the main thread" - Maybe this is true for some particular bitmap functions. But you can use SharpDX.WIC to load bitmaps on any thread. DirectX allows resource creation on any thread. Except if you need to do any ImmediateContext work such as for constructing a texture array, you'll have to do that work on your rendering thread.

**Comment by jtorjo** [comment]
> "[winrt] Loading bitmaps only on the main thread" - Maybe this is true for some particular bitmap functions. But you can use SharpDX.WIC to load bitmaps on any thread. DirectX allows resource creation on any thread. Except if you need to do any ImmediateContext work such as for constructing a texture array, you'll have to do that work on your rendering thread.

Please give me ONE bitmap class where the above it not true.

Clearly, DirectX allows resource creation on any thread - unlike the "great" winrt. I don't see any possible way to load a bitmap that I would then use with win2d, other than on the main thread. If you know of a way, please share it.

(side-note: the only "shortcut" that I had found was to use DDS files - but in my tests, DDS files trim too much information of the original bitmaps, so showing them on the screen would end up being horrible)

I've looked a bit on SharpDX.WIC - and using that would mean I would be tied to SharpDX, while I want to use win2d.

**Comment by Gavin-Williams** [comment]
I'll try to put something together for you and make sure it's actually possible to get it into Win2d.

And a note on DDS, I used to use that format all the time, and you can have uncompressed DDS which looks as good as bitmaps. These days I use PNG's because of there not being even basic support in Windows (like viewing DDS and and icon generation in Explorer) - And I'm aware of the NVidia plugins, but I'm just bored with Microsoft's lack of support for their own format.

It does seem odd though that WinRT has to load on the main thread. That feels like an arbitrary restriction to me. Perhaps someone at Microsoft can look into why it's restricted like that? 

And I also 100% agree that files should be loadable synchronously. An example scenario..

UIThread -------------------------------->
BackgroundLoader --------------------->

There is no need for asynchronous load, we are already on background and the scene can't start or change until assets are loaded and constructed ready for use. I feel like giving us the ability to control which thread the work is done on in conjunction with the existing asynchronous functions gives us greater flexibility in how we conceptualize and implement our solutions.

**Comment by jtorjo** [comment]
> I'll try to put something together for you and make sure it's actually possible to get it into Win2d.
Thanks, look forward to it! I only hope it's actually possible!

> 
> Edit: What about the Win2d bitmap loading functions?
In my testing, the work decent only if loading from the main thread. Long story short, I tried using that, and if you don't do it from the main thread, the time penalty is obscene - I would wait like 5-6 SECONDS for a bitmap to load.

> 
> And a note on DDS, I used to use that format all the time, and you can have uncompressed DDS which looks as good as bitmaps. These days I use PNG's because of there not being even basic support in Windows (like viewing DDS and and icon generation in Explorer) - And I'm aware of the NVidia plugins, but I'm just bored with Microsoft's lack of support for their own format.

Yeah, DDS is a weird beast. Uncompressed DDS is worse than PNG - basically, I use 1280x720 files - PNG - roughly 1Mb, Uncompressed DDS - 3.7Mb (which is pretty much a .bmp bitmap).

win2d can only work with BC1/BC2/BC3 compression formats, and in those formats, the compressed picture looks too different from the original

> 
> It does seem odd though that WinRT has to load on the main thread. That feels like an arbitrary restriction to me. Perhaps someone at Microsoft can look into why it's restricted like that?

"odd" is not the word I'd use :)
**Well, it would be great if someone at Microsoft could look into that.** 

This just shows why a lot of people simply turn away from UWP - it's got sooo many useless restrictions, and most of them come from WinRT. And I just don't understand why Microsoft doesn't seem to care...

> 
> And I also 100% agree that files should be loadable synchronously. An example scenario..

I totally agree. The LoadAsync functions are giving me the creeps.

> ...I feel like giving us the ability to control which thread the work is done on in conjunction with the existing asynchronous functions gives us greater flexibility in how we conceptualize and implement our solutions.

Couldn't agree more!

**Comment by jesbis** [comment]
I see you got some replies on the Q&A page and @Felix-Dev was spot on, but just to reiterate:

We definitely do watch Feedback Hub and that's the best current option, though I agree it often isn't obvious that the feedback is actually seen. We're working on how to better accept and acknowledge feedback on other non-UI WinRT stuff, but we want to make sure it's a sustainable process so that takes time.

----

Regarding threading:

The generic file & storage APIs are out of scope for WinUI/Xaml, but for Xaml we do have an internal backlog issue to consider making changes to [`Image`](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Image) and/or [`BitmapImage`](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Media.Imaging.BitmapImage) to relax their current UI thread affinity and either make their methods more thread-safe or even potentially enable explicitly creating them on background threads. If that's something that would be of interest then I'd recommend opening a new proposal issue in this repo with any data you can provide on your scenario.

**Comment by jtorjo** [comment]
> We definitely do watch Feedback Hub and that's the best current option, though I agree it often isn't obvious that the feedback is actually seen. We're working on how to better accept and acknowledge feedback on other non-UI WinRT stuff, but we want to make sure it's a sustainable process so that takes time.

Right, looking forward to when you'll streamline the process!
My biggest pet peeve now is this: https://docs.microsoft.com/answers/questions/608/please-either-allow-systemio-all-over-hdd-or-drast.html

Everything else - I've figured out in one way or another. Since I can now load SoftwareBitmaps in non-UI threads, I'm happy :)

> 
> Regarding threading:
> 
> The generic file & storage APIs are out of scope for WinUI/Xaml, but for Xaml we do have an internal backlog issue to consider making changes to [`Image`](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Image) and/or [`BitmapImage`](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Media.Imaging.BitmapImage) to relax their current UI thread affinity and either make their methods more thread-safe or even potentially enable explicitly creating them on background threads. If that's something that would be of interest then I'd recommend opening a new proposal issue in this repo with any data you can provide on your scenario.

I'm not very interested in Image/BitmapImage, since those are pretty much wrappers over an Image, and once loaded, there's not much else you can do than show them.

I am interested in WriteableBitmap class - I would really like to be able to load/save/modify it in non UI threads. Right now, my attempts to access it in non UI threads have all thrown exceptions.

So, all in all, i will create a new proposal in a bit.

**Comment by jesbis** [comment]
It's more complex but another currently available option for multithreaded image loading and drawing might be to use `IWicBitmapDecoder` to load images into DirectX texture(s), e.g. using `ID2D1RenderTarget::CreateBitmapFromWicBitmap`.

You can then efficiently draw Direct2D surfaces in Xaml UI trees using one or more `SurfaceImageSources` (or `VirtualSurfaceImageSource` if the images are huge), particularly with `ISurfaceImageSourceNativeWithD2D`, which enables a) multithreaded drawing and b) Direct2D batching across multiple surfaces per frame.

You might also be able to simplify at least some of that using Win2D.

**Comment by jtorjo** [comment]
Yeah, unfortunately Win2D doesn't expose `CreateBitmapFromWicBitmap` - but I'll create a request.
It's imperative to me to be able to do things synchronously, since when it's my turn to draw, I can't load the image right then, `LoadAsync()` APIs are simply useless.

**Comment by jevansaks** [comment]
> Also, I'm curious about where both Winrt/UWP are heading - is there **any** development on them? Are you listening to **any** feedback from developers trying to use it?

This is obviously an area for improvement, but one that I'm passionate about us doing better on. It's going to take some time.

Windows APIs at large have the same problem. It's just that a lot of those old Win32 APIs have been around so long that there are lots of experts outside Microsoft that can answer questions about them just as well as folks inside. :)

To start with, we will start promoting the Q&A site (the docs.microsoft.com/answers site) more as the place to ask these kinds of questions when the area you're asking about isn't backed by a GitHub repo (and most of them aren't). 

I appreciate you raising specific topics as ones that you're having trouble getting answers on or that you're getting unsatisfactory answers on -- I can follow up on those and try to get the owners to jump in.

**Comment by jtorjo** [comment]
Hi @jevansaks ,

Thanks for taking the time to reply!

> This is obviously an area for improvement, but one that I'm passionate about us doing better on. It's going to take some time.

I get that and I'm fine with that - as long as we know where to ask, and what to expect.

I'd really like a place where people could see what's happening to WinRT - stuff Microsoft is working on, and ETAs - and clearly, in that same place, people would ask/vote for the things they want.

As far as I understand (please correct me if I'm wrong), UWP is relying on WinRT for a lot of its API - so if WinRT doesn't support something, nor will UWP. And the clearest example are (lack of) synchronous functions for dealing with loading/saving bitmaps - everything coming from WinRT is asynchronous, and that's how UWP has it. The same is with StorageFile/Folder APIs.

> Windows APIs at large have the same problem. It's just that a lot of those old Win32 APIs have been around so long that there are lots of experts outside Microsoft that can answer questions about them just as well as folks inside. :)

Right, it's just that I'm waaaay more comfoable with Win APIs :D 

> To start with, we will start promoting the Q&A site (the docs.microsoft.com/answers site) more as the place to ask these kinds of questions when the area you're asking about isn't backed by a GitHub repo (and most of them aren't).

Right, where I posted my original question.

> I appreciate you raising specific topics as ones that you're having trouble getting answers on or that you're getting unsatisfactory answers on -- I can follow up on those and try to get the owners to jump in.

Thanks! My biggest pet peeve right now is this - https://docs.microsoft.com/answers/questions/608/please-either-allow-systemio-all-over-hdd-or-drast.html

Another issue - once again, I have no idea where to post this - is that, this to me sounds weird (so I wouldn't use another word :) ) - my UWP app requests broad filesystem access - the user installs it and is **aware** of it, and yet, when my app runs, this is disabled by default. So, I need to tell the user - "Hey, go to Windows Settings, select that <blabla> page, and give my application access".

About LoadAsync() - **I really hate those APIs - all of them**. Sure, it's ok for simple cases. but when doing anything non trivial, you need synchronous access - if I want to do something asyncrhonously, I can  easily span it to another task - just do `Task.Run(() => { ... })` - like I did [here](https://docs.microsoft.com/answers/questions/1994/uwpc-is-there-a-way-to-load-a-bitmap-not-on-the-ma.html) (see my latest answer).

My point is:
a. When things happen async, I can't really know for sure how much time they really take
b. Debugging is a pain
c. the code needs to change, since everything calling an async function needs to become async. which can end up being A LOT. Not to mention constructors and/or properties.

I want to have the choice myself - whether I do something Async() or not - I'm a big boy.

About my last original question - I was wrong, there's a way to load bitmaps not on the main thread. I was pointed out in the Q&A forum, and I'm really really happy for this, since my code became pretty much 5 times faster.
